server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
                     }
                     indexShard.recoverLocallyUpToGlobalCheckpoint(ActionListener.assertOnce(l));
                 })
+                // peer recovery can consume a lot of disk space, so it's worth cleaning up locally ahead of the attempt
+                // operation runs only if the previous operation succeeded, and returns the previous operation's result.
+                // Failures at this stage aren't fatal, we can attempt to recover and then clean up again at the end. #104473
+                .andThenApply(startingSeqNo -> {
+                    Store store = indexShard.store();
+                    store.incRef();
+                    try {
+                        store.cleanupAndVerify("cleanup before peer recovery", indexShard.snapshotStoreMetadata());
+                    } catch (IOException e) {
+                        // if we can't read a snapshot, then we don't have enough intact data locally to create one.
+                        // Make an attempt to clean up temporary files from the previous run. We could also just delete
+                        // everything in the directory but in that case we might want to be a little more thoughtful about
+                        // the exact exception triggered here. Corrupt indices might be recoverable without a full transfer, for
+                        // example.
+                        logger.info(() -> format("cleanup before peer recovery failed on shard [{}]", indexShard.shardId()), e);
+                        try {
+                            recoveryTarget.cleanTempFiles();
```

Comment:
should we hold a ref to `recoveryTarget` for this call?

Reply:
probably yes, I'll add that.